### PR TITLE
Unit testing framework for caffe2_ref_test

### PR DESCRIPTION
**Pull request by ezyang**
This adds the necessary infrastructure for performing per-node
unit tests for various operators.

I will add more tests in a later commit.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: tests/caffe2_ref_test.py [hunk 1]
```diff
@@ -27,6 +27,109 @@
...
+        self.inputs = inputs
+
+    def run(self, test_self):
+        # TODO: In some cases we should generate multiple random inputs
+        # and test (ala Hypothesis)
+        args = create_input(self.inputs)
+        np_results = self.np_impl(*args)
+        if not isinstance(np_results, tuple):
+            np_results = (np_results, )
+        input_names = list(map(lambda i: "input_{}".format(i), range(len(args))))
```
Comment (bddppq) [inline tests/caffe2_ref_test.py]: nit: use list comprehension? that way you don't need to worry whether map returning a list or a generator.

File: onnx_caffe2/backend.py [hunk 2]
```diff
@@ -152,6 +153,7 @@ class Caffe2Backend(Backend):
         'Reshape': '_create_reshape',
         'GlobalAveragePool': '_create_global_pool_op',
         'GlobalMaxPool': '_create_global_pool_op',
+        'Dot': '_create_dot',
```
Comment (bddppq) [inline onnx_caffe2/backend.py]: remove this?

File: tests/caffe2_ref_test.py [hunk 3]
```diff
@@ -27,6 +27,109 @@
...
+        output_names = list(map(lambda i: "output_{}".format(i), range(len(np_results))))
+        node_def = helper.make_node(self.node.name, input_names, output_names, **self.node.kwargs)
+        caffe2_results = c2.run_node(node_def, args)
+        for i in range(len(output_names)):
```
Comment (bddppq) [inline tests/caffe2_ref_test.py]: It's better to also check the number of outputs matches.

**Comment by ezyang** [comment]
Forced pushed with CR fixes.

**Pull request merged**